rfcs/0086-create-secret.md

```diff
@@ -0,0 +1,64 @@
+---
+feature: New Syntax Create Secret
+authors:
+  - "TabVersion"
+start_date: "2024/03/18"
+---
+
+# Introducing SECRETs for Secure Storage of Sensitive Information
+
+## Background
+
+This RFC proposes a new syntax and implementation for securely storing and referencing sensitive information,
+such as database passwords and API keys, in SQL statements using a SECRET catalog.
+
+The primary goals are to enhance the security of production environments by encrypting sensitive data, applying access controls, 
+and preventing abuse of API keys by developers.
+
+## Proposed Syntax
+
+```sql
+-- create secret
+create secret <secret-name> from <expr>
+```
+
+We always store SECRETs as bytea type, which maximally ensures that the information we store can adapt to different usage requirements.
+
+```sql
+-- use secret
+create table t ( ... )
+with (
+    ...
+    properties.sasl.password = s-<secret-name>
+)
+```
+
+### Usage Scenarios
+
+There are two main scenarios for using SECRETs:
+
+* Direct value usage (e.g., `properties.sasl.password`): The SECRET needs to be correctly decrypted in the meta and passed along with the table fragment to the connector.
+* File path usage (e.g., `properties.ssl.certificate.location`): The Kafka SDK requires a file path, so a temporary file needs to be written on the CN, and the path is passed to the Kafka SDK.
+
+## Implementation
+
+SECRETs are always stored as bytea type in SQL to ensure flexibility in adapting to different usage requirements.
+If a user needs to use a certificate, its contents are also stored as bytea type and distributed to Compute Nodes (CNs) as needed.
```

Comment:
So you will just store them as plaintext on disk? If so, we might need to use some encryption function provided by the DB, otherwise, storing credential in plaintext seems unacceptable to me.

Reply:
To prevent plaintext in data at rest, either way works for me:

1. Run a symmetrical encryption & decryption algorithm in RW's code. 
2. Rely on the DB's encryption

In both approaches, the encryption key will be managed by another system (such as K8s) and inject to RW/DB via config file/env var/etc.

Reply:
prefer 2. Let's ask @yezizp2012

Reply:
Considering the need to be compatible with encryption implementations of different DB backends, I am more inclined prefer 1. 🤔

Reply:
I prefer 1. I guess it'd be hard to implement encryption for every DB backends that we support. Think about sqlite.

Reply:
It's okay to tell users data-at-rest encryption is not available for SQLite.

I also tend to 1 for PG meta-store backend. While if Vault is in use, we can just let Vault do it.